metadata/moa/scripts/nbconverted/3.align-moa.py

```diff
@@ -0,0 +1,531 @@
+        .reset_index(drop=True)
+        .loc[:, ["InChIKey14", "broad_id_unique"]]
+        .drop_duplicates()
+        .assign(broad_date=broad_id_col)
+        .dropna()
+        .reset_index(drop=True)
+    )
+
+    return map_long_df
+
+
+# ## Step 1: Define InChIKey14-to-Broad ID Map
+
+# In[4]:
+
+
+mapping_col = "InChIKey14"
+
+repurposing_col = "broad_id_20170327"
+broad_id_other_dates = ["20170327", "20180516", "20180907", "20200324"]
+
+
+# In[5]:
+
+
+# Load platemap data
+file = pathlib.PurePath("../platemaps/broad_sample_info.tsv")
+all_platemap_df = pd.read_csv(file, sep="\t")
+
+print(all_platemap_df.shape)
+all_platemap_df.head()
+
+
+# In[6]:
+
+
+# How many unique broad IDs?
+len(all_platemap_df.broad_id.unique())
+
+
+# In[7]:
+
+
+# Load MOA map
+map_file = pathlib.PurePath("clue/broad_id_map.csv")
+map_df = pd.read_csv(map_file)
+
+print(map_df.shape)
+map_df.head()
+
+
+# In[8]:
+
+
+# Create MOA map long mapping every Broad ID to InChIKey14
+all_broad_map = []
+for date in broad_id_other_dates:
+    broad_id_col = f"broad_id_{date}"
+    deprecated_broad_id_col = f"deprecated_broad_id_{date}"
+
+    select_cols = [mapping_col, broad_id_col, deprecated_broad_id_col]
+    map_subset_df = map_df.loc[:, select_cols]
+
+    broad_long_df = split_pipe_broad_id(map_subset_df, broad_id_col)
+    all_broad_map.append(broad_long_df)
+
+    if date != "20170327":
+        deprecated_broad_long_df = split_pipe_broad_id(
+            map_subset_df, deprecated_broad_id_col
+        )
+        all_broad_map.append(deprecated_broad_long_df)
+
+# Combine full broad_id map
+all_broad_map = (
+    pd.concat(all_broad_map)
+    # This step drops InChIKeys and Broad_ID combos that exist in multiple dates
+    .drop_duplicates(subset=[mapping_col, "broad_id_unique"]).reset_index(drop=True)
+)
+
+print(all_broad_map.shape)
+all_broad_map.head()
+
+
+# In[9]:
+
+
+# Are there Duplicate InChIKey14s?
+inchi_counts = all_broad_map.loc[:, mapping_col].squeeze().value_counts()
+inchi_counts.hist()
+
+
+# In[10]:
+
+
+# Show an example of the most egregious multi-mapping InChiKey14
+all_broad_map.loc[all_broad_map.loc[:, mapping_col] == inchi_counts.index[0], :]
+
+
+# ## Step 2: Perform the Merge with the Platemap Data
+# 
+# This step identifies the InChIKey14 map for the Cell Painting Repurposing Hub Data Subset.
+
+# In[11]:
+
+
+merged_platemap_df = (
+    all_platemap_df.merge(
+        all_broad_map, left_on="broad_id", right_on="broad_id_unique", how="left"
+    )
+    # We don't care about plate info here
+    .drop_duplicates(subset=["broad_id", mapping_col]).loc[
+        :, ["broad_sample", "broad_id", mapping_col]
+    ]
+    # Drop DMSO
+    .dropna(subset=["broad_sample"])
+)
+
+print(merged_platemap_df.shape)
+merged_platemap_df.head()
+
+
+# In[12]:
+
+
+# Are there Duplicate Broad IDs
+broad_id_counts = merged_platemap_df.broad_id.value_counts()
+broad_id_counts.hist()
+
+
+# In[13]:
+
+
+# There are indeed duplicated broad IDs!
```

Comment:
I was surprised to see duplicates. You probably haven't spent time figuring this out but do you think this due to the same `deprecated_broad_id` and `broad_id` mapping to two different `InChIKey14`?

Reply:
This is a great question, and one reason code review is so powerful! There must be some reason why there are duplicates... Indeed, sometimes it is because of deprecation, _but this is not always the case_. I added the `broad_date` column back to this and saw this (as an example, there are others in this list too):

![image](https://user-images.githubusercontent.com/7340421/80253028-80811f80-8646-11ea-9529-4119842bb3a3.png)

Basically, it looks like `BRD-K50691590`, `BRD-K98733875`, and `BRD-K72922393` are duplicated b/c of a deprecation, while `BRD-K68103045` and `BRD-K28936863` are duplicated for some other reason.

I think this matters because we should not try to resolve `broad_ids` with mutiple moa/target information in the instances where we see it map to a deprecated broad_id (this is slightly different than your question of duplicate broad IDs with different InChIkeys, but the question made me take the idea a bit further). This happens exactly once! 

In the next commit I will remove this compound from alternative moa/target resolution. It seems like when there is non-deprecated moa/target info, it is good enough and we can remove the deprecated info. Thanks! :taco:

Reply:
Turns out the offending `broad_id` with the deprecation had identical moa/target info as another alternate. Adding a `.drop_duplicates(["broad_id", "moa", "target"])` call cleared this issue up. Without this fix, one `broad_id` would have redundant moa/target info in the `alternative` column as in the original column.

I also will keep the `broad_date` column in the map so that future users will know exactly which resource provided the map.

Reply:
Awesome sleuthing! :) :taco: to you! 

And a good call to keep the `broad_date` column in the final map.